### PR TITLE
KEP-3962: Mutating admission policy documentation

### PR DESCRIPTION
**Pull request by Jefftree**
### Description

This documents KEP-3962 for the alpha release of the enhancement in 1.32

Supersedes https://github.com/kubernetes/website/pull/48467

Comments still outstanding from that PR are addressed.

### CODE REVIEW

**Comment by sftim** [comment]
> Comments still outstanding from that PR that need to be addressed here.

/retitle [WIP] KEP-3962: Mutating admission policy documentation

Sounds like this is not yet ready for review, so I'll mark it as a draft. @Jefftree if you can get it ready, that's helpful.

**Review by jpbetz — APPROVED** [review]
/approve
/lgtm
For technical content for alpha

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (tengqm) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:54]:
```suggestion
Mutating admission policy is part of the cluster control-plane. You should write
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (tengqm) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:56]:
```suggestion
experiment with mutating admission policy.
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (tengqm) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:44]:
```suggestion
  MutatingAdmissionPolicy objects specify what kind they are expecting for their parameter resource.
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (Jefftree) [reply content/en/docs/reference/access-authn-authz/mutating-admission-policy.md]: removed

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (Jefftree) [reply content/en/docs/reference/access-authn-authz/mutating-admission-policy.md]: Fixed. Is the title fine or should that also be lower case?

`## Getting Started with Mutating Admission Policy`

**Review by Jefftree — COMMENTED** [review]
Comments addressed @tengqm

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:52]: Based on our style guide, I recommend:
```suggestion
## Getting Started with MutatingAdmissionPolicies
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:258]:
```suggestion
The same mutation can be written as a [JSON Patch](https://jsonpatch.com/) as follows:
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:156]: Please use a `code_sample` here.
See https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#source-code-files

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:260]: Please use a `code_sample` here.
See https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#source-code-files

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:62]: Please use a `code_sample` here.
See https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#source-code-files

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:79]: Easier / more readable to use a YAML multiline string here, IMO.

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:173]: Easier / more readable to use a YAML multiline string here, IMO.

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:151]:
```suggestion
Parameter resources allow a policy configuration to be separate from its
definition. A policy can define `paramKind`, which outlines GVK of the parameter
resource, and then a policy binding ties a policy by name (via `policyName`) to a
particular parameter resource via `paramRef`.
```

At beta, we should explain what a GVK is. OK to add that clarification for alpha too.

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:124]: A Pod would typically need a `.metadata.name` - can we either mark this as incomplete, or add in what's missing?

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:113]:
```suggestion
You can test that admission policy by creating a Pod:
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:43]:
```suggestion
  Parameter resources are always in that the Kubernetes API (they can be built-in types or extensions,
  such as a {{< glossary_tooltip term_id="CustomResourceDefinition" text="CustomResourceDefinition" >}} (CRD).
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:24]: (nit) An Oxford-like comma may help:
```suggestion
[server side apply merge strategy](/docs/reference/using-api/server-side-apply/#merge-strategy),
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:100]: Is this true?
```suggestion
The `.spec.mutations` field consists of a list of expressions that evaluate to resource patches.
The emitted patches may be either [apply configurations](#applyconfiguration) or [JSON Patch](#json-patch)
patches. You cannot specify an empty list of mutations.
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:131]: Let's mark the `.metadata` as elided.

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:211]:
```suggestion
The parameter resource could be the following object:
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:208]:
```suggestion
    namespace: default # Avoid using the "default" namespace outside of tutorial examples
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:254]: Optionally, and if correct:
```suggestion

The `apiVersion`, `kind`, `metadata.name` and `metadata.generateName` are always accessible from the root of the
object. No other metadata properties are accessible.
This means that you cannot emit an apply configuration to make any change to an object's labels or annotations.
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:299]:
```suggestion
Each evaluated `expression` must return an array of `JSONPatch` values. The
`JSONPatch` type represents one operation from a JSON patch.
```

(would be easier to have named the type as `JSONPatchOperation`, but too late now I suspect)

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:310]:
```suggestion
To define a JSON object for a patch operation `value`, use CEL `Object` types. For example:
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:322]:
```suggestion
To use strings containing '/' and '~' as JSONPatch path keys, use `jsonpatch.escapeKey()`. For example:
```

**Review by sftim — COMMENTED** [review]
Thanks. It's good to have this doc ready for review on schedule.

We try to avoid putting tutorials and conceptual explanations into the docs reference section, so I'd love to see changes to this PR.

Please consider rewording https://kubernetes.io/docs/concepts/policy/#apply-policies-using-validatingadmissionpolicy to mention mutating admission policies too. This change is important.

For beta (and ideally even for alpha), we should (as in **could**, not must) turn this guide-style explanation into:
- a tutorial that helps people try out VAP and MAP in their toy cluster
  - yes, I mean **one combined tutorial** that teaches both kinds of admission policies
  - yes, I know that ValidatingAdmissionPolicy is GA and MutatingAdmissionPolicy is alpha; that's OK
- a shorter reference explanation of MutatingAdmissionPolicy, to live within the reference pages

Please consider doing or supporting that work eventually, @Jefftree. "Supporting" need not mean you do the work; it could be more around putting out a call for help.
As an opinion, but also the opinion of a former tech lead, that refactoring to a tutorial would really help people learn about K8s admission.

I've added some inline feedback too. See what you think.

---

Here's the part of the page I'd keep for the reference section (perhaps slightly expanded):

> CEL expressions have access to the types needed to create JSON patches and objects:
> 
> - `JSONPatch` - CEL type of JSON Patch operations. JSONPatch has the fields `op`, `from`, `path` and `value`.
>   See [JSON patch](https://jsonpatch.com/) for more details. The `value` field may be set to any of: string,
>   integer, array, map or object.  If set, the `path` and `from` fields must be set to a
>   [JSON pointer](https://datatracker.ietf.org/doc/html/rfc6901/) string, where the `jsonpatch.escapeKey()` CEL
>   function may be used to escape path keys containing `/` and `~`.
> - `Object` - CEL type of the resource object.
> - `Object.<fieldName>` - CEL type of object field (such as `Object.spec`)
> - `Object.<fieldName1>.<fieldName2>...<fieldNameN>` - CEL type of nested field (such as `Object.spec.containers`)
> 
> CEL expressions have access to the contents of the API request, organized into CEL variables as well as some other useful variables:
> 
> - `object` - The object from the incoming request. The value is null for DELETE requests.
> - `oldObject` - The existing object. The value is null for CREATE requests.
> - `request` - Attributes of the API request.
> - `params` - Parameter resource referred to by the policy binding being evaluated. Only populated if the policy has a ParamKind.
> - `namespaceObject` - The namespace object that the incoming object belongs to. The value is null for cluster-scoped resources.
> - `variables` - Map of composited variables, from its name to its lazily evaluated value.
>   For example, a variable named `foo` can be accessed as `variables.foo`.
> - `authorizer` - A CEL Authorizer. May be used to perform authorization checks for the principal (user or service account) of the request.
>   See https://pkg.go.dev/k8s.io/apiserver/pkg/cel/library#Authz
> - `authorizer.requestResource` - A CEL ResourceCheck constructed from the `authorizer` and configured with the
>   request resource.

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:38]:
```suggestion
  If you only want to set an `owner` label for Pods, and not other API kinds, the binding is where you
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:41]:
```suggestion
- A _parameter resource_ provides information to a MutatingAdmissionPolicy to make it a concrete
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:196]: (nit / fix not required for alpha)
```suggestion
MutatingAdmissionPolicyBinding that uses a cluster-wide parameter; the same parameter
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [reply content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:211]: This particular parameter is a Sidecar, and Sidecar's API looks like an object (has `.metadata` and `.spec`, has `.metadata.name`, etc).

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:226]: (nit)
```suggestion
#### `ApplyConfiguration` {#patch-type-apply-configuration}
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:256]:
```suggestion
#### `JSONPatch` {#patch-type-json-patch}
```

**Review by sftim — COMMENTED** [review]
A bit more feedback

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:218]:
```suggestion
  namespace: default # Avoid using the "default" namespace outside of tutorial examples
```

**Comment by sftim** [comment]
Overall: I'm wary to approve this as-is, but @Jefftree once you have read the feedback, please feel free to make the minimum set of changes you feel are right for alpha stage. Then request a new review (you could consider the auto-assigned reviewed, @natalisucks and @reylejano).

**Comment by AnshumanTripathi** [comment]
Hi @jpbetz 👋 !
I'm reaching out from the Docs team. Just checking in as we approach Docs Freeze on Tuesday November 26th 18:00 PDT. This documentation appears to still be under review. To meet the Docs Freeze, this PR must have a technical review as well as lgtm and approve labels applied, without any unaddressed comments or concerns from SIG Docs. Thank you!

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (Jefftree) [reply content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:124]: added name

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (Jefftree) [reply content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:79]: I agree for longer expressions but this seems reasonably contained, now that it's moved into a separate file

**Comment by Jefftree** [comment]
@sftim comments addressed. Tried to put together a minimal set of references with a single example for applyconfiguration and json patch without too many examples making this into a tutorial. As a follow up (after docs freeze), will create a new page as a tutorial for VAP and MAP.

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (tengqm) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:37]: What does "the above resources" refer to?

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (tengqm) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:137]: Is it able to check the labels of an object?
I don't mean I want to change it.

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (Jefftree) [reply content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:37]: Reordered, should be the mutatingadmissionpolicy along with any parameter resources.

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (Jefftree) [reply content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:137]: Next line: "This means that you cannot emit an apply configuration to make any change to an object's labels or annotations."

Comment (Jefftree) [reply content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:138]: yes, thanks for catching!

File: content/en/docs/reference/command-line-tools-reference/feature-gates/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates/mutating-admission-policy.md:11]:
```suggestion
  - stage: alpha 
    defaultValue: false
    fromVersion: "1.30"
```

The existing data are correct.

File: content/en/docs/reference/command-line-tools-reference/feature-gates/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates/mutating-admission-policy.md:14]:
```suggestion
Enable [MutatingAdmissionPolicy](/docs/reference/access-authn-authz/mutating-admission-policy/) support for [CEL](https://kubernetes.io/docs/reference/using-api/cel/) mutations be used in admission control.

For Kubernetes v1.30 and v1.31, this feature gate existed but has no effect.
```

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:12]: I think we'll want to manually specify this.
```suggestion
{{< feature-state for_k8s_version="v1.32" state="alpha" >}}
<!-- due to feature gate history, use manual version specification here -->
```

**Comment by Jefftree** [comment]
@sftim comments addressed, thanks!

File: content/en/docs/reference/access-authn-authz/mutating-admission-policy.md
Comment (sftim) [reply content/en/docs/reference/access-authn-authz/mutating-admission-policy.md:139]: OK, also: annotations?

**Review by sftim — COMMENTED** [review]
/lgtm

with a [quibble](https://github.com/kubernetes/website/pull/48646/files#r1858984851)

**Pull request merged**
